python-checks/src/main/java/org/sonar/python/checks/SpecialMethodReturnTypeCheck.java

```diff
@@ -0,0 +1,361 @@
+        // Hence, we abort in this case.
+
+        return;
+      }
+    }
+
+    if (numReturnedExpressions != 2) {
+      addIssueOnReturnedExpressions(ctx, returnStatement, String.format(INVALID_GETNEWARGSEX_ELEMENT_COUNT_MESSAGE, numReturnedExpressions));
+      return;
+    }
+
+    // Exactly two expressions are being returned
+    final Expression firstElement = returnedExpressions.get(0);
+    final Expression secondElement = returnedExpressions.get(1);
+
+    if (!firstElement.type().canBeOrExtend(BuiltinTypes.TUPLE) ||
+      !secondElement.type().canBeOrExtend(BuiltinTypes.DICT)) {
+      ctx.addIssue(firstElement.firstToken(), secondElement.lastToken(), INVALID_GETNEWARGSEX_TUPLE_MESSAGE);
+    }
+  }
+
+  private static InferredType getReturnStmtType(ReturnStatement returnStatement) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      return InferredTypes.NONE;
+    }
+
+    if (returnedExpressions.size() == 1) {
+      return returnedExpressions.get(0).type();
+    }
+
+    return InferredTypes.TUPLE;
+  }
+
+  /**
+   * Calls {@code ctx.addIssue} for a return statement such that...
+   *
+   * ...all returned expressions are marked as the source of the issue if the return statement contains such expressions
+   * ...the return keyword is marked as the source of the issue if the return statement does not contain any expressions
+   */
+  private static void addIssueOnReturnedExpressions(SubscriptionContext ctx, ReturnStatement returnStatement, String message) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      ctx.addIssue(returnStatement.returnKeyword(), message);
+    } else {
+      final Token firstExpressionToken = returnedExpressions.get(0).firstToken();
+      final Token lastExpressionToken = returnedExpressions.get(returnedExpressions.size() - 1).lastToken();
+
+      ctx.addIssue(firstExpressionToken, lastExpressionToken, message);
+    }
+  }
+
+  private static ReturnStmtCollector collectReturnStmts(FunctionDef funDef) {
+    final ReturnStmtCollector returnExpressionCollector = new ReturnStmtCollector();
+    funDef.body().accept(returnExpressionCollector);
+
+    return returnExpressionCollector;
+  }
+
+  private static class ReturnStmtCollector extends BaseTreeVisitor {
+    private final List<ReturnStatement> returnStmts = new ArrayList<>();
+    private List<Token> yieldKeywords = new ArrayList<>();
+    private boolean raisesWhitelistingException = false;
+
+    public List<ReturnStatement> getReturnStmts() {
+      return returnStmts;
+    }
+
+    public List<Token> getYieldKeywords() {
+      return yieldKeywords;
+    }
+
+    public boolean raisesWhitelistingException() {
+      return raisesWhitelistingException;
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatement pyReturnStatementTree) {
```

Comment:
Minor: can you please remove the "py" prefix from parameter names? I know the suggestion comes from the interface, however, I don't think it really adds information.